devlib/instrument/gem5stats.py

```diff
@@ -0,0 +1,152 @@
+#    Copyright 2017 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import division
+import csv
+import re
+
+from devlib.instrument import Instrument, CONTINUOUS, MeasurementsCsv
+from devlib.exception import TargetError, HostError
+
+class Gem5StatsInstrument(Instrument):
+    '''
+    Instrument enabling power monitoring in Gem5
+
+    This instrument parses the stats file produced by Gem5 and extracts power
+    values. More information on the "stats.txt" file structure can be found 
+    here:
+    
+    http://pages.cs.wisc.edu/~david/courses/cs752/Spring2015/gem5-tutorial/part1/gem5_stats.html
+    
+    This instrument also implements Gem5's ROIs (Regions of Interest). A ROI is
+    an annotation on a time interval marked by start and end cursors. This enables
+    power analysis on a specific time window which spans over the execution of
+    a workload for example. Up to 8 ROIs can be used, with or without 
+    overlapping.
+    '''
+
+    mode = CONTINUOUS
+    dump_head='---------- Begin Simulation Statistics ----------'
+    dump_tail='---------- End Simulation Statistics   ----------'
+    
+    def __init__(self, target, channel_map, stats_file_path):
+        super(Gem5StatsInstrument, self).__init__(target)
+        
```

Comment:
Since this instrument works *only* with gem5, there should be a check here to ensure that the target's platform is indeed a gem5 platform and raise a `ValueError` here if it isn't.